File: backend/app/api/v1/recipes.py
from fastapi import APIRouter, Query, UploadFile

from app.dependencies import (
    CurrentUser,
    ImageServiceDep,
    OptionalUser,
    RecipeServiceDep,
    ScraperServiceDep,
    SessionDep,
)
from app.schemas.recipe import (
    ImageResponse,
    PaginatedResponse,
    RecipeCreate,
    RecipeImportRequest,
    RecipeResponse,
    RecipeUpdate,
)

router = APIRouter(prefix="/recipes", tags=["recipes"])


@router.get("", response_model=PaginatedResponse)
async def list_recipes(
    recipe_service: RecipeServiceDep,
    user: OptionalUser,
    page: int = Query(1, ge=1),
    per_page: int = Query(20, ge=1, le=100),
    search: str | None = None,
    tag_ids: list[str] | None = Query(None),
    category_id: str | None = None,
    is_favorite: bool | None = None,
    min_rating: int | None = Query(None, ge=1, le=5),
    max_cook_time: int | None = Query(None, ge=0),
    sort_by: str = Query(
        "created_at",
        pattern="^(title|created_at|rating|cook_time_minutes)$",
    ),
    sort_order: str = Query("desc", pattern="^(asc|desc)$"),
) -> PaginatedResponse:
    return await recipe_service.list_recipes(
        page=page,
        per_page=per_page,
        search=search,
        tag_ids=tag_ids,
        category_id=category_id,
        is_favorite=is_favorite,
        min_rating=min_rating,
        max_cook_time=max_cook_time,
        sort_by=sort_by,
        sort_order=sort_order,
        public_only=user is None,
    )


@router.get("/{slug}", response_model=RecipeResponse)
async def get_recipe(
    slug: str,
    recipe_service: RecipeServiceDep,
    user: OptionalUser,
) -> RecipeResponse:
    return await recipe_service.get_recipe_by_slug(slug, public_only=user is None)


@router.post("", response_model=RecipeResponse, status_code=201)
async def create_recipe(
    data: RecipeCreate,
    recipe_service: RecipeServiceDep,
    current_user: CurrentUser,
    session: SessionDep,
) -> RecipeResponse:
    result = await recipe_service.create_recipe(data, current_user.id)
    await session.commit()
    return result


@router.put("/{recipe_id}", response_model=RecipeResponse)
async def update_recipe(
    recipe_id: str,
    data: RecipeUpdate,
    recipe_service: RecipeServiceDep,
    current_user: CurrentUser,
    session: SessionDep,
) -> RecipeResponse:
    result = await recipe_service.update_recipe(recipe_id, data)
    await session.commit()
    return result


@router.delete("/{recipe_id}", status_code=204)
async def delete_recipe(
    recipe_id: str,
    recipe_service: RecipeServiceDep,
    current_user: CurrentUser,
    session: SessionDep,
) -> None:
    await recipe_service.delete_recipe(recipe_id)
    await session.commit()


@router.post("/import", response_model=RecipeResponse, status_code=201)
async def import_recipe(
    data: RecipeImportRequest,
    scraper_service: ScraperServiceDep,
    current_user: CurrentUser,
    session: SessionDep,
) -> RecipeResponse:
    recipe = await scraper_service.import_from_url(data.url, current_user.id)
    await session.commit()
    return RecipeResponse.model_validate(recipe)


@router.patch("/{recipe_id}/favorite", response_model=RecipeResponse)
async def toggle_favorite(
    recipe_id: str,
    recipe_service: RecipeServiceDep,
    current_user: CurrentUser,
    session: SessionDep,
) -> RecipeResponse:
    result = await recipe_service.toggle_favorite(recipe_id)
    await session.commit()
    return result


@router.patch("/{recipe_id}/rating", response_model=RecipeResponse)
async def set_rating(
    recipe_id: str,
    rating: int | None = Query(None, ge=1, le=5),
    recipe_service: RecipeServiceDep = None,
    current_user: CurrentUser = None,
    session: SessionDep = None,
) -> RecipeResponse:
    result = await recipe_service.set_rating(recipe_id, rating)
    await session.commit()
    return result


@router.post("/{recipe_id}/images", response_model=ImageResponse, status_code=201)
async def upload_image(
    recipe_id: str,
    file: UploadFile,
    image_service: ImageServiceDep,
    current_user: CurrentUser,
    session: SessionDep,
    is_primary: bool = False,
    alt_text: str | None = None,
) -> ImageResponse:
    result = await image_service.upload_image(recipe_id, file, is_primary, alt_text)
    await session.commit()
    return ImageResponse.model_validate(result)


@router.delete("/images/{image_id}", status_code=204)
async def delete_image(
    image_id: str,
    image_service: ImageServiceDep,
    current_user: CurrentUser,
    session: SessionDep,
) -> None:
    await image_service.delete_image(image_id)
    await session.commit()

File: frontend/src/api/recipes.ts
import api from "./client";
import type {
  PaginatedResponse,
  Recipe,
  RecipeCreateData,
  Tag,
} from "./types";

interface ListRecipesParams {
  page?: number;
  per_page?: number;
  search?: string;
  tag_ids?: string[];
  is_favorite?: boolean;
  min_rating?: number;
  max_cook_time?: number;
  sort_by?: string;
  sort_order?: string;
}

export async function listRecipes(
  params: ListRecipesParams = {}
): Promise<PaginatedResponse> {
  const searchParams = new URLSearchParams();
  if (params.page) searchParams.set("page", String(params.page));
  if (params.per_page) searchParams.set("per_page", String(params.per_page));
  if (params.search) searchParams.set("search", params.search);
  if (params.is_favorite !== undefined)
    searchParams.set("is_favorite", String(params.is_favorite));
  if (params.min_rating) searchParams.set("min_rating", String(params.min_rating));
  if (params.max_cook_time)
    searchParams.set("max_cook_time", String(params.max_cook_time));
  if (params.sort_by) searchParams.set("sort_by", params.sort_by);
  if (params.sort_order) searchParams.set("sort_order", params.sort_order);
  if (params.tag_ids) {
    params.tag_ids.forEach((id) => searchParams.append("tag_ids", id));
  }

  return api.get("recipes", { searchParams }).json();
}

export async function getRecipe(slug: string): Promise<Recipe> {
  return api.get(`recipes/${slug}`).json();
}

export async function createRecipe(data: RecipeCreateData): Promise<Recipe> {
  return api.post("recipes", { json: data }).json();
}

export async function updateRecipe(
  id: string,
  data: Partial<RecipeCreateData>
): Promise<Recipe> {
  return api.put(`recipes/${id}`, { json: data }).json();
}

export async function deleteRecipe(id: string): Promise<void> {
  await api.delete(`recipes/${id}`);
}

export async function importRecipe(url: string): Promise<Recipe> {
  return api.post("recipes/import", { json: { url } }).json();
}

export async function toggleFavorite(id: string): Promise<Recipe> {
  return api.patch(`recipes/${id}/favorite`).json();
}

export async function setRating(
  id: string,
  rating: number | null
): Promise<Recipe> {
  const searchParams = rating !== null ? { rating: String(rating) } : {};
  return api.patch(`recipes/${id}/rating`, { searchParams }).json();
}

export async function listTags(): Promise<Tag[]> {
  return api.get("tags").json();
}

export async function createTag(name: string): Promise<Tag> {
  return api.post("tags", { json: { name } }).json();
}

export async function deleteTag(id: string): Promise<void> {
  await api.delete(`tags/${id}`);
}

export async function uploadImage(
  recipeId: string,
  file: File,
  isPrimary: boolean = false
): Promise<void> {
  const formData = new FormData();
  formData.append("file", file);
  await api.post(`recipes/${recipeId}/images`, {
    body: formData,
    searchParams: { is_primary: String(isPrimary) },
  });
}
